pd/src/app.rs

```diff
@@ -443,24 +452,63 @@ impl App {
 
                 let mut next_rates = Vec::new();
                 let mut next_validator_statuses = Vec::new();
+
+                // this is a bit complicated: because we're in the EndBlock phase, and the
+                // delegations in this block have not yet been committed, we have to combine
+                // the delegations in pending_block with the ones already committed to the
+                // state. otherwise the delegations committed in the epoch threshold block
+                // would be lost.
+                let mut committed_delegation_changes =
+                    state.delegation_changes(prev_epoch.index).await?;
+                for (id_key, delta) in &pending_block.lock().unwrap().delegation_changes {
+                    let _ = *committed_delegation_changes
+                        .entry(id_key.clone())
+                        .and_modify(|change| *change += delta)
+                        .or_insert(*delta);
```

Comment:
I think this might be a bit clearer if it were written as

```suggestion
                    *committed_delegation_changes
                        .entry(id_key.clone())
                        .or_insert(0) += delta;
```

(assuming I'm reading the logic right) -- this way, there's only one code path to consider (adding delta), and it's more explicit that the default value is 0.